### PR TITLE
Asset dirs

### PR DESCRIPTION
**Pull request by bpoldrack**

### CODE REVIEW

File: onyo/commands/tests/test_mkdir.py [hunk 1]
```diff
@@ -153,25 +153,6 @@ def test_dir_exists(repo: OnyoRepo, directory: str) -> None:
     assert repo.git.is_clean_worktree()
 
 
-@pytest.mark.repo_files(*directories)  # used as files to test errors
-@pytest.mark.parametrize('file', directories)
-def test_dir_exists_as_file(repo: OnyoRepo, file: str) -> None:
```
Comment (TobiasKadelka) [inline onyo/commands/tests/test_mkdir.py:158]: I think this test (and similar CLI tests later) should not be deleted.

Instead, the assertions just need minimal adjustments, e.g. not erroring anymore, and e.g.
`assert Path(file).is_file()` -> `assert Path(file).is_directory()`.

File: onyo/lib/tests/test_commands_mkdir.py [hunk 2]
```diff
@@ -18,13 +17,6 @@ def test_onyo_mkdir_errors(inventory: Inventory) -> None:
                   dirs=[dir_path],
                   message="some subject\n\nAnd a body")
 
-    # mkdir on existing asset path
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_mkdir.py:21]: I would additionally add here a test that verifies onyo errors correctly when calling `onyo mkdir` on a directory which is already an asset directory :)

Related to #565

File: onyo/lib/tests/test_commands_mv.py [hunk 3]
```diff
@@ -36,14 +36,6 @@ def test_onyo_mv_errors(inventory: Inventory) -> None:
                   destination=dir_path / "doesnotexist" / "somewhere",
                   message="some subject\n\nAnd a body")
 
-    # move directory to existing file
-    pytest.raises(ValueError,
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_mv.py:40]: Instead, this should be a test that checks for an error if `onyo mv` is tried to be used to rename an asset dir, which is illegal.

Related to #562

File: onyo/lib/tests/test_commands_mv.py [hunk 4]
```diff
@@ -239,3 +231,32 @@ def test_onyo_mv_and_rename(inventory: Inventory) -> None:
...
+
+    onyo_mv(inventory,
+            source=source,
+            destination=asset_path,
+            message="Move asset into asset to create asset dir")
+    assert inventory.repo.git.get_hexsha('HEAD~1') == old_hexsha
+    assert inventory.repo.git.is_clean_worktree()
+    assert not source.exists()
+    # Turned destination into asset dir:
+    assert inventory.repo.is_asset_dir(asset_path)
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_mv.py:260]: Nitpick: we might here want to test both, the caches and the filesystem, so as to be sure that both are updated through the change.

File: onyo/lib/tests/test_commands_new.py [hunk 5]
```diff
@@ -255,3 +260,18 @@ def test_onyo_new_clones(inventory: Inventory) -> None:
     # equals existing asset except for path and serial:
     assert all(v == new_asset[k] for k, v in existing_asset.items() if k not in ['serial', 'path'])
     assert new_asset['serial'] == 'whatever'
+
+
+@pytest.mark.ui({'yes': True})
+def test_onyo_new_asset_dir(inventory: Inventory) -> None:
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_new.py:266]: Based on #561, we should add a test with `"is_asset_directory": False`, and possibly `"is_asset_directory": false`, too, that create a normal asset.

I assume that #561 is just happening for the CLI because everything gets interpreted as a string instead of a bool, but to make sure the value to the key is considered correctly is IMO still worth doing. 

What happens (or should happen) for things like `"is_asset_directory": ["this fullfills truthyness requirements"]`? > depends obviously on the check done, but since this is about a reserved key, it is worth considering :)

File: onyo/commands/tests/test_mkdir.py (hunk 1 above)
Comment (bpoldrack) [reply onyo/commands/tests/test_mkdir.py:158]: This should be tested in the python tests, IIRC.
I'd like to not repeat functionality tests in the CLI tests, but rather just test the wrapper does the correct wrapping :)

File: onyo/lib/tests/test_commands_set.py [hunk 6]
```diff
@@ -466,82 +341,44 @@ def test_onyo_set_depth(inventory: Inventory) -> None:
...
 
 
-@pytest.mark.repo_contents(
-    ["one_that_exists.test", "type: one\nmake: that\nmodel: exists\nserial: test\nsome_key: value"])
 @pytest.mark.ui({'yes': True})
-def test_onyo_set_depth_zero(inventory: Inventory) -> None:
-    """Calling `onyo_set(depth=0)` is legal and selects
-    all assets from all subpaths."""
-    key_value = {"this_key": "that_value"}
+def test_onyo_set_asset_dir(inventory: Inventory) -> None:
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_set.py:345]: Maybe this happens already and I just overlooked it (the diff is a bit weird to read at times).

There should be another explicit test that verifies that `onyo_set()` turns an asset into an asset dir.

File: onyo/lib/tests/test_commands_unset.py [hunk 7]
```diff
@@ -70,24 +60,6 @@ def test_onyo_unset_errors(inventory: Inventory) -> None:
                   message="some subject\n\nAnd a body")
 
 
-@pytest.mark.ui({'yes': True})
-def test_onyo_unset_on_empty_directory(inventory: Inventory) -> None:
```
Comment (TobiasKadelka) [inline onyo/lib/tests/test_commands_unset.py:74]: Should probably  add a test above that verifies that this type of call now errors on directories.

**Review by TobiasKadelka — COMMENTED** [review]
I really like this one, thanks a lot for the PR! :)

Two things about my comments:

1. The changes are a lot, and the diff is for some files weird for me. I might have requested some additions, which are already there, just not were I looked for them. I hope you can forgive me when this happens :)
2. I know that not everything can/should be tested in this PR into every corner case, but I still added comments requesting some additional tests for scenarios that I consider relevant (e.g. likely to be done), or where I found bugs for, and I think these will be easy to add/fix. Whenever this is not as easy as I think, please add an issue for them, so that they can be added later!

File: onyo/lib/tests/test_commands_mv.py (hunk 4 above)
Comment (bpoldrack) [reply onyo/lib/tests/test_commands_mv.py:260]: That's not necessary here. We test the command did `commit`. Cache update only needs to be done when committing, so this is included in `commit` and tested for with the respective unit tests.

Comment (bpoldrack) [reply onyo/lib/tests/test_commands_mv.py:260]: Furthermore the `repo.is_xxx` methods are in fact using the cache.

File: onyo/lib/tests/test_commands_new.py (hunk 5 above)
Comment (bpoldrack) [reply onyo/lib/tests/test_commands_new.py:266]: There's a confusion here. What is `"is_asset_directory": false` supposed to be? That's not valid python ;-)

As you suspected, it's only about strings coming in from the CLI and being converted correctly before handed off to us by argparse.

Comment (bpoldrack) [reply onyo/lib/tests/test_commands_new.py:266]: (Fixed already)

File: onyo/lib/tests/test_commands_set.py (hunk 6 above)
Comment (bpoldrack) [reply onyo/lib/tests/test_commands_set.py:345]: Lines 367ff and 377ff do that (both directions).

File: onyo/lib/tests/test_commands_mv.py (hunk 3 above)
Comment (bpoldrack) [reply onyo/lib/tests/test_commands_mv.py:40]: Done.

**Review by aqw — APPROVED** [review]
There's no way I can review this in-depth right now. But superficially it looks good to me. And it works.

What's important is that this works and gets in before the hackathon.

Afterwards we will immediately expand test coverage, and gain more confidence that it's actually doing the right thing. ;-)
